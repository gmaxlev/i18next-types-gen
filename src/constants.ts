export const DEFAULT_INTERFACE_NAME = "LocalesKeys";
