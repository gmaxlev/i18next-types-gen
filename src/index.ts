export { genLocalesType } from "./gen";
export { genLocalesTypeIntoFile } from "./fs";
