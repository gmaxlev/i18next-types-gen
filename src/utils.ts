import { isRecord } from "tsguarder";
import fsp from "fs/promises";

/**
 * Recursively extract keys from a JSON object.
 * @param value The value to extract keys from.
 * @param set The set to add keys to.
 * @param from The path to the current value.
 */
export function extractKeys(
  value: unknown,
  set = new Set<string>(),
  from: string[] = []
) {
  if (isRecord(value)) {
    for (const key in value) {
      extractKeys(value[key], set, [...from, key]);
    }
  } else {
    set.add(from.join("."));
  }

  return set;
}

/**
 * Create an interface definition from a set of keys.
 * @param set The set of keys.
 * @param name The name of the interface.
 */
export function createInterfaceDefinition(set: Set<string>, name: string) {
  let string = `export interface ${name} {`;
  set.forEach((key) => {
    const formatNamespace = key.replace(".", ":");
    string += `\n  '${formatNamespace}': string;`;
  });
  string += "\n}\n";
  return string;
}

/**
 * Reads a locale file and returns the parsed JSON.
 * @param path The path to the locale file.
 */
export async function readLocaleFile(path: string) {
  let buffer: Buffer;

  try {
    buffer = await fsp.readFile(path);
  } catch (e) {
    console.error("Failed to read file: " + path);
    throw e;
  }

  let parsed: unknown;

  try {
    parsed = JSON.parse(buffer.toString());
  } catch (e) {
    console.error("Failed to parse JSON file: " + path);
    throw e;
  }

  try {
    isRecord.assert(parsed, path);
  } catch (e) {
    console.error("Invalid JSON file: " + path);
    throw e;
  }

  return parsed;
}

/**
 * Joins multiple locales into a single object.
 * @param sources The sources to join.
 */
export function joinLocales(sources: Record<PropertyKey, unknown>[]) {
  let common: Record<PropertyKey, unknown> = {};
  sources.forEach((source) => {
    common = {
      ...common,
      ...source,
    };
  });
  return common;
}
