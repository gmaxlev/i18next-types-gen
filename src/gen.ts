import { isArray, isRecord, isString, isUndefined } from "tsguarder";
import {
  createInterfaceDefinition,
  extractKeys,
  joinLocales,
  readLocaleFile,
} from "./utils";
import { DEFAULT_INTERFACE_NAME } from "./constants";

export interface GenLocalesTypeOptions {
  sources: string[];
  interfaceName?: string | undefined;
}

export async function genLocalesType(options: GenLocalesTypeOptions) {
  try {
    isRecord.assert(options, "options");

    const { sources, interfaceName = DEFAULT_INTERFACE_NAME } = options;

    isArray.assert(sources, "sources");

    if (!isUndefined(interfaceName)) {
      isString.assert(interfaceName, "interfaceName");
    }

    const contents = await Promise.all(
      sources.map((source, index) => {
        isString.assert(source, `sources[${index}]`);
        return readLocaleFile(source);
      })
    );

    const common = joinLocales(contents);

    const keys = extractKeys(common);

    const typeDefinition = createInterfaceDefinition(
      keys,
      interfaceName ?? "LocalesKeys"
    );

    return {
      typeDefinition,
      size: keys.size,
    };
  } catch (e) {
    console.error("Failed to generate type definition");
    throw e;
  }
}
