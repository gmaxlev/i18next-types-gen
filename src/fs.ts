import { isRecord, isString } from "tsguarder";
import path from "path";
import fsp from "fs/promises";
import { genLocalesType, GenLocalesTypeOptions } from "./gen";

interface GenLocalesTypeIntoFileOptions extends GenLocalesTypeOptions {
  output: string;
}

export async function genLocalesTypeIntoFile(
  options: GenLocalesTypeIntoFileOptions
) {
  try {
    console.log("‚òÑÔ∏è Start generating locales type definition");

    isRecord.assert(options, "options");

    const { sources, interfaceName, output } = options;

    const { typeDefinition, size } = await genLocalesType({
      sources,
      interfaceName,
    });

    console.log(`üåà Generated ${size} keys`);

    isString.assert(options.output, "output");

    const dirname = path.dirname(output);

    try {
      await fsp.mkdir(dirname, { recursive: true });
    } catch (e) {
      console.error("Failed to create directory: " + dirname);
      throw e;
    }

    try {
      await fsp.writeFile(output, typeDefinition);
    } catch (e) {
      console.error("Failed to write file: " + output);
      throw e;
    }
    console.log("üéâ Saved locales type definition into file:", output);
  } catch (e) {
    console.error("Failed to generate type definition into file");
    throw e;
  }
}
