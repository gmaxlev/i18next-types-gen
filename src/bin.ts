import { Command } from "commander";
import { isRecord, isString, isUndefined } from "tsguarder";
import pkg from "../package.json";
import { genLocalesTypeIntoFile } from "./fs";

const program = new Command();

let description = `CLI to ${pkg.name} package\n`;
description += `See docs at ${pkg.homepage}`;

program.name(pkg.name).description(description).version(pkg.version);

program
  .command("run")
  .description("Generate locales type definition into a file")
  .argument("<sources>", "sources paths")
  .argument("<output>", "output path")
  .option("-n, --name <name>", "interface name")
  .action(async (sourcesString, output, options) => {
    isString.assert(sourcesString, "sources");
    isString.assert(output, "sources");
    isRecord.assert(options, "options");

    const sources = sourcesString.split(",");

    const name = "name" in options ? options["name"] : undefined;

    if (!isUndefined(name)) {
      isString.assert(name, "name");
    }

    await genLocalesTypeIntoFile({
      sources,
      output,
      interfaceName: name,
    });
  });

program.parse();
