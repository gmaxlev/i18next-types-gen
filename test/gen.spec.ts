import { genLocalesType } from "../src";
import { vol } from "memfs";

jest.mock("fs/promises");

describe("gen.ts", () => {
  afterEach(() => {
    vol.reset();
  });

  const fsSnapshot = {
    "file1.json": JSON.stringify({
      a: {
        b: "c",
      },
    }),
    "file2.json": JSON.stringify({
      d: {
        e: "f",
      },
    }),
    "file3.json": JSON.stringify({
      g: "h",
    }),
  };

  describe("genLocalesType()", () => {
    test("Should generate type definition and use specified name", async () => {
      vol.fromNestedJSON(fsSnapshot, "/src");

      const definition = await genLocalesType({
        sources: ["/src/file1.json", "/src/file2.json", "/src/file3.json"],
        interfaceName: "SpecialLocalesKeys",
      });

      expect(definition).toMatchSnapshot();
    });

    test("Should generate type definition and use the default interface name", async () => {
      vol.fromNestedJSON(fsSnapshot, "/src");

      const definition = await genLocalesType({
        sources: ["/src/file1.json", "/src/file2.json", "/src/file3.json"],
      });

      expect(definition).toMatchSnapshot();
    });
  });
});
