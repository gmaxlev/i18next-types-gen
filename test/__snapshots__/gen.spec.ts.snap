// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gen.ts genLocalesType() Should generate type definition and use specified name 1`] = `
{
  "size": 3,
  "typeDefinition": "export interface SpecialLocalesKeys {
  'a:b': string;
  'd:e': string;
  'g': string;
}
",
}
`;

exports[`gen.ts genLocalesType() Should generate type definition and use the default interface name 1`] = `
{
  "size": 3,
  "typeDefinition": "export interface LocalesKeys {
  'a:b': string;
  'd:e': string;
  'g': string;
}
",
}
`;
